### ISSUE TITLE
android shield gamepad buttons R2 and L2 do nothing

### ISSUE THREAD

**Issue by ictbram** [godotengine/godot#12696]
With Godot 2.1.4, 
android app export on Nvidia shield console (Tegra X1),
android 7.0

The included gamepad of the shield console does not seem to trigger events with the L2 and R2 buttons.
I do not get any input events while pressing them with the standard gamepads on that device.
Those buttons do trigger events with an xbox 360 gamepad connected to that device or on another device from nvidia, the shield portable, it works as expected.
They work in almost every other setup, computers applications, html, xbox one, firetv (first gen) 

I followed this tutorial for remote debugging:
http://codetuto.com/2016/06/godot-engine-remote-debug-live-edit/ 
but It can't get it to work, deploying works, but no debugging output, so I just printed things to the screen

Here are prints from a computer with that xbox 360 gamepad connected when I press those buttons:
"Device 0 ID 1019 Event: JoystickButton Pressed: 1 Index: 6 pressure 0" and "Device 0 ID 1024 Event: JoystickButton Pressed: 0 Index: 6 pressure 0" for L2
"Device 0 ID 1335 Event: JoystickButton Pressed: 1 Index: 7 pressure 0" and "Device 0 ID 1338 Event: JoystickButton Pressed: 0 Index: 7 pressure 0" for R2

the code that prints the input event is:
func input(event):
	if event.type == InputEvent.JOYSTICKBUTTON:
		Utils.debugPrint(str(event))

I've made a really small project that just prints the gamepad input events the same way
[gamepadTester.zip](https://github.com/godotengine/godot/files/1445704/gamepadTester.zip)
, same result

Nothing gets printed when I press L2 and R2 on the nvidia shield console, even without the if event.type == InputEvent.JOYSTICKBUTTON:
I've posted my question on godotengine.org qa and it was suggested to post my question here
https://godotengine.org/qa/19307/android-shield-gamepad-buttons-r2-en-l2-do-nothing

**Comment by kubecz3k** [comment]
cc @Hinsbart

**Comment by akien-mga** [comment]
Is this still reproducible in the current master branch / latest Godot 3.2 beta 5 builds?

If so, can you retrieve the GUID of the affected gamepad with `Input.get_joy_guid(0)` (assuming it's the device 0)?

**Comment by psiker** [comment]
Yes, I tested with Godot 3.2 beta 5
printing Input.get_joy_guid(0) in func _input(event): 
gave me 4e564944494120436f72617649
I got 3 prints when I pressed L2
no prints when I pressed R2

**Comment by akien-mga** [comment]
It seems like there are indeed no mapping for this GUI in SDL2 for Android (see https://github.com/gabomdq/SDL_GameControllerDB).

If you are able to extract the mapping for this controller on an Android device via SDL2, you could add them to this repository and we'll get them too in our next update. The process doesn't seem well documented for Android though... It sounds like you'd need to build `test/controllermap.c` for Android somehow. https://github.com/gabomdq/SDL_GameControllerDB#mapping-tools

**Comment by KoBeWi** [comment]
Can anyone still reproduce this bug in Godot 3.2.3 or any later release?

**Comment by ictbram** [comment]
Yes, exactly like three years back. 
I’ve just made a new project in Godot 3.2.3 with a button, textfield and copy pasted my simple test script.  L2 and R2 trigger on the portable shield and on the shield TV with an xbox 360 gamepad, but not with the supplied nvidia gamepad on the shield TV.
The shield TV device has the latest updates, shows Shield Android TV SW Version 8.2.1(32.8.435.31)

**Comment by michael-conrad** [comment]
> Yes, exactly like three years back.
> I’ve just made a new project in Godot 3.2.3 with a button, textfield and copy pasted my simple test script. L2 and R2 trigger on the portable shield and on the shield TV with an xbox 360 gamepad, but not with the supplied nvidia gamepad on the shield TV.
> The shield TV device has the latest updates, shows Shield Android TV SW Version 8.2.1(32.8.435.31)

I'm testing 3.2.4rc1 now on my shield tv and have bizarre issues happening with the left analog stick.

I press up, it goes left, down, it goes right. It gets stuck going left....

Are the analog sticks working correctly for you?

I'm a bit surprised at this behavior as the nvidia shield controller pretty much maps the same as an xbox controller.

**Comment by ictbram** [comment]
Yes, the same for me with the shield controller on the shield, the left stick gives left direction when pressing to the right, also keeps giving left signals after that. I recently linked a ps4 controller with my shield device, using that controller, everything works as it should.

**Comment by michael-conrad** [comment]
Is there way to force register a GUID as a different GUID for a gamepad, or to load an SDL mapping at runtime for a specific GUID?

**Comment by michael-conrad** [comment]
> Yes, I tested with Godot 3.2 beta 5
> printing Input.get_joy_guid(0) in func _input(event):
> gave me 4e564944494120436f72617649
> I got 3 prints when I pressed L2
> no prints when I pressed R2

I found this entry in the gamecontrolldb.txt

4e564944494120436f72706f72617469,NVIDIA Controller,a:b0,b:b1,dpdown:h0.4,dpleft:h0.8,dpright:h0.2,dpup:h0.1,leftshoulder:b9,leftstick:b7,lefttrigger:a4,leftx:a0,lefty:a1,rightshoulder:b10,rightstick:b8,righttrigger:a5,rightx:a2,righty:a3,start:b6,x:b2,y:b3,platform:Android,

which seems to match the Xbox series:

050000005e040000130b0000ffff3f00,Xbox Series Controller,a:b0,b:b1,back:b4,dpdown:b12,dpleft:b13,dpright:b14,dpup:b11,guide:b5,leftshoulder:b9,leftstick:b7,lefttrigger:a4,leftx:a0,lefty:a1,rightshoulder:b10,rightstick:b8,righttrigger:a5,rightx:a2,righty:a3,start:b6,x:b2,y:b3,platform:Android,

**Comment by michael-conrad** [comment]
I've tried:

```
Input.add_joy_mapping("4e564944494120436f72706f72617469,NVIDIA Controller,leftx:a0,lefty:a1,dpdown:h0.4,rightstick:b8,rightshoulder:b10,rightx:a2,start:b6,righty:a3,dpleft:h0.8,lefttrigger:a4,x:b2,dpup:h0.1,back:b4,leftstick:b7,leftshoulder:b9,y:b3,a:b0,dpright:h0.2,righttrigger:a5,b:b1,platform:Android,", true)
```

but it seems to *ignore* this?

**Comment by Calinou** [comment]
> or to load an SDL mapping at runtime for a specific GUID?

Yes, you can set the `SDL_GAMECONTROLLERCONFIG` environment variable before running Godot:

```bash
# Linux/macOS
SDL_GAMECONTROLLERCONFIG="config;string;here" path/to/godot.x86_64

# Windows
set SDL_GAMECONTROLLERCONFIG=config;string;here
path/to/godot.exe
```

**Comment by michael-conrad** [comment]
Ok, 

This is not making any sense to me.

I did:

```gdscript
func _ready():
	Input.remove_joy_mapping("4e564944494120436f72706f72617469")
	Input.add_joy_mapping("4e564944494120436f72706f72617469,Shield-TV Controller,leftx:a0,lefty:a1,dpdown:h0.4,rightstick:b8,rightshoulder:b10,rightx:a2,start:b6,righty:a3,dpleft:h0.8,lefttrigger:a4,x:b2,dpup:h0.1,back:b4,leftstick:b7,leftshoulder:b9,y:b3,a:b0,dpright:h0.2,righttrigger:a5,b:b1,platform:Android,", true)
```

And the behavior is the same. I even tried with just removing the mapping. No change in behavior.

**Comment by michael-conrad** [comment]
> > or to load an SDL mapping at runtime for a specific GUID?
> 
> Yes, you can set the `SDL_GAMECONTROLLERCONFIG` environment variable before running Godot:
> 
> ```shell
> # Linux/macOS
> SDL_GAMECONTROLLERCONFIG="config:string:here" path/to/godot.x86_64
> ```
> 
> ```batchfile
> :: Windows
> set SDL_GAMECONTROLLERCONFIG=config:string:here
> path/to/godot.exe
> ```

Unfortunately we are dealing with an Android device. (NVIDIA Shield TV).

**Comment by michael-conrad** [comment]
Whatever the issue is with NVIDIA controllers, it also impacts my bluetooth OUYA controllers.

I used a test controller mapping where I swapped the axis for leftx and lefty and that sort of worked, there is still the issue where it thinks the direction selected on the left stick isn't going back to center.

**Comment by michael-conrad** [comment]
I think this ticket should be renamed to reflect it is a general "joypad controller" not working correctly for NVIDIA Shield TV - at least for NVIDIA controller and OUYA controller

**Comment by michael-conrad** [comment]
logcat output - just in case if helps.

```bash
02-02 23:39:16.462 28914 28933 I godot   : Godot Engine v3.2.4.rc1.official - https://godotengine.org
02-02 23:39:16.470 28914 28933 I godot   : OpenGL ES 2.0 Renderer: NVIDIA Tegra
02-02 23:39:16.615 28914 28933 I godot   : OpenGL ES Batching: ON
02-02 23:39:16.650 28914 28933 I godot   :  
02-02 23:39:19.267 28914 28933 I godot   : Joypad: joy_connection_changed(0, True)
02-02 23:39:19.267 28914 28933 I godot   : Joypad: 4e564944494120436f72706f72617469, Custom Mapping
02-02 23:39:19.267 28914 28933 I godot   : Joypad: 0, 4e564944494120436f72706f72617469, Custom Mapping
02-02 23:39:19.267 28914 28933 E godot   : **ERROR**: Condition "!joy_names.has(p_device)" is true. Returned: ""
02-02 23:39:19.267 28914 28933 E godot   :    At: main/input_default.cpp:1254:get_joy_guid_remapped() - Condition "!joy_names.has(p_device)" is true. Returned: ""
02-02 23:39:19.267 28914 28933 E godot   : **ERROR**: Condition "!joy_names.has(p_device)" is true. Returned: ""
02-02 23:39:19.267 28914 28933 E godot   :    At: main/input_default.cpp:1254:get_joy_guid_remapped() - Condition "!joy_names.has(p_device)" is true. Returned: ""
...
```

**Comment by Calinou** [comment]
@michael-conrad You can call [`Input.add_joy_mapping()`](https://docs.godotengine.org/en/latest/classes/class_input.html#class-input-method-add-joy-mapping) in your main menu script's `_ready()` function.

**Comment by michael-conrad** [comment]
> > Unfortunately we are dealing with an Android device. (NVIDIA Shield TV).
> 
> You can call [`Input.add_joy_mapping()`](https://docs.godotengine.org/en/latest/classes/class_input.html#class-input-method-add-joy-mapping) in your main menu script's `_ready()` function. This should let you add a mapping at run-time on any platform including Android. Make sure to set the `update_existing` argument of `Input.add_joy_mapping()` to `true` when calling the method.

As noted previously, this does not fix the issue with NVIDIA Shield TV's and gamepads.

The behavior looks more like something low level is going very wrong, as I get the same behavior with the OUYA bluetooth controller. Though an x-box controller hooked up via USB works fine.

**Comment by michael-conrad** [comment]
I downloaded the joypad demo from the assets store and added a mapping for NVIDIA controller.

Here are screenshots of joypad inputs showing badness with all control on physical devices at rest.

![image](https://user-images.githubusercontent.com/5229492/106913394-f2a48b80-66d1-11eb-8d43-f55df4435a76.png)

![image](https://user-images.githubusercontent.com/5229492/106913440-fd5f2080-66d1-11eb-9f80-57d8ad85653f.png)

I added the add joy mapping in _ready()

```gdscript
func _ready():
	set_physics_process(true)
	Input.connect("joy_connection_changed", self, "_on_joy_connection_changed")
	Input.add_joy_mapping("4e564944494120436f72706f72617469,NVIDIA Xbox Style Controller,a:b0,b:b1,back:b15,dpdown:h0.4,dpleft:h0.8,dpright:h0.2,dpup:h0.1,guide:b5,leftshoulder:b9,leftstick:b7,lefttrigger:a4,leftx:a0,lefty:a1,rightshoulder:b10,rightstick:b8,righttrigger:a5,rightx:a2,righty:a3,start:b6,x:b2,y:b3,platform:Android,", true)
```

I also changed the behavior so that the active device being displayed is auto switched to any device added. (Saves from having to use keyboard/mouse on TV).

```gdscript
func _on_joy_connection_changed(device_id, connected):
	if device_id == cur_joy:
		joypad_name.set_text(Input.get_joy_name(device_id) if connected else "")
	if connected:
		joypad_number.value=device_id
```

There is something very much wrong with the wireless controllers on android-tv. I have an x-box 360 USB controller, and it works fine.

This pretty much kills godot for android tv. At least for the nvidia shield.

**Comment by michael-conrad** [comment]
See #45712

**Comment by michael-conrad** [comment]
> Yes, the same for me with the shield controller on the shield, the left stick gives left direction when pressing to the right, also keeps giving left signals after that. I recently linked a ps4 controller with my shield device, using that controller, everything works as it should.

I've created two pull requests to fix the issue, (#45771, #45772).  Seems to be a cross up between the virtual mouse and the gamepad.

**Comment by akien-mga** [comment]
Fixed by #45771.
